scripts/broadcast-mndp-discover.nse

```diff
@@ -0,0 +1,249 @@
+-- |   MAC Address: 00:0c:29:6d:a7:63, IP Address: fe80::20c:29ff:fe6d:a763; Identity: MikroTik; Version: 6.42.12 (long-term); Platform: MikroTik; Software ID: GXCE-KYGV; Uptime: 1h14m; Board: x86; Unpacking: None; Interface: ether1
+-- |   MAC Address: 00:0c:29:8b:de:c6, IP Address: 10.1.1.123; Identity: MikroTik; Version: 6.10; Platform: MikroTik; Software ID: 33UY-8JI2; Uptime: 0h42m; Board: x86; Unpacking: None; Interface: ether1
+-- |_  MAC Address: 00:0c:29:8b:de:c6, IP Address: fe80::20c:29ff:fe8b:dec6; Identity: MikroTik; Version: 6.10; Platform: MikroTik; Software ID: 33UY-8JI2; Uptime: 0h42m; Board: x86; Unpacking: None; Interface: ether1
+--
+-- @args broadcast-mndp-discover.address
+--       address to which the probe packet is sent. (default: 255.255.255.255)
+-- @args broadcast-mndp-discover.timeout
+--       socket timeout (default: 5s)
+---
+
+author = "Brendan Coles"
+license = "Same as Nmap--See https://nmap.org/book/man-legal.html"
+categories = {"discovery", "broadcast", "safe"}
+
+prerule = function() return ( nmap.address_family() == "inet") end
+
+local arg_address = stdnse.get_script_args(SCRIPT_NAME .. ".address")
+local arg_timeout = stdnse.parse_timespec(stdnse.get_script_args(SCRIPT_NAME .. ".timeout"))
+
+-- Listens for MNDP response messages.
+--@param interface Network interface to listen on.
+--@param timeout Time to listen for a response.
+--@param responses table to insert response data into.
+local mndpListen = function(interface, timeout, responses)
+  local condvar = nmap.condvar(responses)
+  local start = nmap.clock_ms()
+  local listener = nmap.new_socket()
+  local filter = 'udp src port 5678 and udp dst port 5678 and src host not ' .. interface.address
+  listener:set_timeout(500)
+  listener:pcap_open(interface.device, 1024, true, filter)
+
+  while (nmap.clock_ms() - start) < timeout do
+    local status, _, _, l3data = listener:pcap_receive()
+
+    if not status then
+      goto continue
+    end
+
+    local p = packet.Packet:new(l3data, #l3data)
+    local data = l3data:sub(p.udp_offset + 9)
+
+    if #data < 4 then
+      goto continue
+    end
+
+    stdnse.print_debug(1, "Received MNDP response from %s (%s bytes)", p.ip_src, string.len(data))
+
+    local tlv_type, tlv_len, tlv_value, pos
+    pos = 1
+    --local header = data:sub(pos, pos + 1)
+    pos = pos + 2
+    --local seqno = data:sub(pos, pos + 1)
+    pos = pos + 2
+
+    local mac_address, identity, version, platform, uptime, software_id, board, unpacking, interface
+    while (pos < #data) do
+      -- TLV Type - Unsigned integer (2 bytes)
+      tlv_type = string.unpack(">I2", data:sub(pos, pos + 1))
+      pos = pos + 2
+
+      -- TLV Length - Unsigned integer (2 bytes)
+      tlv_len  = string.unpack(">I2", data:sub(pos, pos + 1))
+      pos = pos + 2
+
+      -- TLV Value
+      tlv_value = data:sub(pos, pos + tlv_len - 1)
+      pos = pos + tlv_len
```

Comment:
Just to show you a trick how you can do these unpacks with a one-liner:
```lua
tlv_type, tlv_value, pos = string.unpack(">I2s2", data, pos)
-- tlv_len is no longer needed because #tlv_value is the same thing
```

Reply:
Thanks. Done.